### DISCUSSION TITLE
Webpack 5 detection

### DISCUSSION THREAD

**Discussion by ooihuiying** [storybookjs/storybook#18132] [Help] [answered]
Hi, I have a question regarding webpack version detection in storybook.

If i have a project that uses Storybook, specifies to use webpack5 as the core builder, **is there a way for a storybook addon to find out which webpack version is being selected?**

```
// Projects adds this to main.js and specifies to use webpack5
core: {
    builder: 'webpack5',
},

```

In the addon, lets say i have a function that extends Storybook's webpack config as mentioned here https://storybook.js.org/docs/react/configure/webpack. 

```
// Export a function. Accept the base config as the only param.
module.exports = {
  webpackFinal: async (config, { configType }) => {
    // `configType` has a value of 'DEVELOPMENT' or 'PRODUCTION'
    // You can change the configuration based on that.
    // 'PRODUCTION' is used when building the static version of storybook.

    // Make whatever fine-grained changes you need
    config.module.rules.push({
      test: /\.scss$/,
      use: ['style-loader', 'css-loader', 'sass-loader'],
      include: path.resolve(__dirname, '../'),
    });

    // Return the altered config
    return config;
  },
};
```
Is  possible to do something in this function to read the webpack version number being used?

I did some experiment. It seems that i can do something like below to find out the webpack version that is running:

```
let version;
configuration.plugins.forEach((plugin: any) => {
        if (Object.keys(plugin).includes("version")){
            // assuming that the HtmlWebpackPlugin is the one with version field
            version = plugin.version;
        }
    })
```
When webpack5 is chosen as the core builder, the version number of HtmlWebpackplugin (one of the plugin inside `configuration.plugins`) will be 5. If the default webpack4 is used, the version number will be 4. From the doc, it seems that this [plugin](https://storybook.js.org/docs/react/configure/webpack) is used by storybook to generate the preview page.  

However, this approach seems to be hacky. The reason why it's hacky is because 
1. i am assuming that the plugin with a version field is the HtmlWebpackPlugin...this is because, i am not able to derive the plugin's name from configuration.plugins....
2. I am unsure if the HtmlWebpackPlugin will always be present to give me the webpack version.

May i know if anyone can give me advice on this? Thank you.

**Comment by ooihuiying** [comment]
@shilman Thank you.

**Comment by shilman** [comment, marked as answer]
I believe you can do the following:

```
  webpackFinal: async (config, { presets }) => {
    const webpack = await presets.apply('builder');
    console.log({ builder });
  }
```

`builder` will either be `undefined` (webpack4), a string like `'webpack4'` or `'webpack5'` (or, starting in 6.5 `'@storybook/builder-webpack4'`/`'...5'`), or an object `{ name: 'webpack4|...', options: { ... } }`. Hope that helps!

**Reply by ooihuiying** [reply]
Hi @shilman, thank you for your reply.

I tried the above method and it didn't work. apply('builder') seems to give an undefined value. 

However, changing 'builder' to 'core'  worked.
 ```
await pre.apply('core').then((build) => {
        console.log(build); // gives me {builder: 'webpack5'}
    });
```

**Reply by shilman** [reply]
indeed. typo in my suggestion -- glad you figured it out!
